File: 2_add_two_numbers/2_add_two_numbers.js

class ListNode {
    constructor(val, next) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

/**
 * 
 * @param {Array} link 
 * @returns 
 */
function prepareLink(link) {
    let head = new ListNode(link[0]);
    for (let i = 1; i<= link.length - 1; i++) {
        let node = head
        while (node.next) node = node.next;
        node.next = new ListNode(link[i]);
    }
    return head
}

/**
 * 
 * @param {ListNode} head 
 */
function printLinkedList(head) {
    console.log("[ ")
    let ptr = head;
    while (ptr) {
        console.log(ptr.val,",");
        ptr = ptr.next;
    }
    console.log("]");
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
 var addTwoNumbers = function(l1, l2) {
    let head = null;
    let prev = null;
    let temp = null;
    let carry = 0;

    // while tw list exists
    while (l1 !== null || l2 !== null) {
        let l1Val = l1 === null ? 0 : l1.val;
        let l2Val = l2 === null ? 0 : l2.val;

        let sum = carry + l1Val + l2Val;

        // update carry 
        carry = sum >= 10 ? 1 : 0;

        // update sum
        sum = sum < 10 ? sum : sum % 10;

        // create new node wih sum
        temp = new ListNode(sum);

        if (head === null) {
            head = temp;
        } else {
            prev.next = temp;
        }
        prev = temp;

        //  to the next child nodes of l1 and l2

        if (l1 !== null) l1 = l1.next;
        if (l2 !== null) l2 = l2.next;
    }

    if (carry > 0){
        temp.next = new ListNode(carry);
    }

    return head;
};

l1 = prepareLink([2,4,9]);
printLinkedList(l1);

l2 = prepareLink([5,6,4,9]);
printLinkedList(l2);

let sol = addTwoNumbers(l1, l2);
printLinkedList(sol);

File: leetcode/2_add_two_numbers/2_add_two_numbers.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
    def prepare_link(link):
        head = ListNode(link[0])
        for el in link[1:]:
            node = head
            while node.next:
                node = node.next
            node.next = ListNode(el)
        return head

    def print_linked_list(head) -> None:
        print("[", end="")
        ptr = head
        while ptr:
            print(ptr.val, end=", ")
            ptr = ptr.next
        print("]")


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = None
        prev = None
        temp = None
        carry = 0

        # while two list exists
        while (l1 is not None or 
               l2 is not None):
            l1_val = 0 if l1 is None else l1.val
            l2_val = 0 if l2 is None else l2.val

            _sum = carry + l1_val + l2_val

            # update carry
            carry = 1 if _sum >= 10 else 0

            # update sum
            _sum = _sum if _sum < 10 else _sum % 10

            # create new node with sum
            temp = ListNode(_sum)

            if head is None:
                head = temp
            else:      
                prev.next = temp
          
            prev = temp

            # to the next child nodes of l1 and l2
            if l1 is not None:
                l1 = l1.next
            if l2 is not None:
                l2 = l2.next
        
        if (carry > 0):
            temp.next = ListNode(carry)
        
        return head
            
            
        
if __name__ == "__main__":
    l1 = ListNode.prepare_link([2,4,9])
    ListNode.print_linked_list(l1)

    l2 = ListNode.prepare_link([5,6,4,9])
    ListNode.print_linked_list(l2)

    sol = Solution().addTwoNumbers(l1, l2)
    ListNode.print_linked_list(sol)
